components/scream/ekat/cmake/EkatCreateUnitTest.cmake

```diff
@@ -0,0 +1,202 @@
+    set(THREAD_INCREMENT ${RETURN_VAL})
+  endif()
+
+  if (NOT MPI_START_RANK GREATER 0)
+    message (FATAL_ERROR "Error! MPI_START_RANK is <=0.")
+  endif()
+  if (NOT MPI_END_RANK GREATER 0)
+    message (FATAL_ERROR "Error! MPI_END_RANK is <=0.")
+  endif()
+  if (MPI_INCREMENT GREATER 0 AND MPI_START_RANK GREATER MPI_END_RANK)
+    message (FATAL_ERROR "Error! MPI_START_RANK > MPI_END_RANK, but the increment is positive.")
+  endif()
+  if (MPI_INCREMENT LESS 0 AND MPI_START_RANK LESS MPI_END_RANK)
+    message (FATAL_ERROR "Error! MPI_START_RANK < MPI_END_RANK, but the increment is negative.")
+  endif()
+  if (NOT THREAD_START GREATER 0)
+    message (FATAL_ERROR "Error! THREAD_START is <=0.")
+  endif()
+  if (NOT THREAD_END GREATER 0)
+    message (FATAL_ERROR "Error! THREAD_END is <=0.")
+  endif()
+  if (THREAD_INCREMENT GREATER 0 AND THREAD_START GREATER THREAD_END)
+    message (FATAL_ERROR "Error! THREAD_START > THREAD_END, but the increment is positive.")
+  endif()
+  if (THREAD_INCREMENT LESS 0 AND THREAD_START LESS THREAD_END)
+    message (FATAL_ERROR "Error! THREAD_START < THREAD_END, but the increment is negative.")
+  endif()
+
+  # Check both, in case user has negative increment
+  if (MPI_END_RANK GREATER 1 OR MPI_START_RANK GREATER 1)
+    if ("${ecut_MPI_EXEC_NAME}" STREQUAL "")
+      set (ecut_MPI_EXEC_NAME "mpiexec")
+    endif()
+    if ("${ecut_MPI_NP_FLAG}" STREQUAL "")
+      set (ecut_MPI_NP_FLAG "-n")
+    endif()
+  endif()
+
+  set(CURR_RANKS ${MPI_START_RANK})
+  set(CURR_THREADS ${THREAD_START})
+
+  if (ecut_EXE_ARGS)
+    set(invokeExec "./${target_name} ${ecut_EXE_ARGS}")
+  else()
+    set(invokeExec "./${target_name}")
+  endif()
+
+  while (NOT CURR_RANKS GREATER ${MPI_END_RANK})
+    while (NOT CURR_THREADS GREATER ${THREAD_END})
+      # Create the test
+      set(FULL_TEST_NAME ${target_name}_ut_np${CURR_RANKS}_omp${CURR_THREADS})
+      if (${CURR_RANKS} GREATER 1)
+        add_test(NAME ${FULL_TEST_NAME}
+                 COMMAND sh -c "${ecut_MPI_EXEC_NAME} ${ecut_MPI_NP_FLAG} ${CURR_RANKS} ${ecut_MPI_EXTRA_ARGS} ${invokeExec}")
+      else()
+        add_test(NAME ${FULL_TEST_NAME}
+                 COMMAND sh -c "${invokeExec}")
+      endif()
+      math(EXPR CURR_CORES "${CURR_RANKS}*${CURR_THREADS}")
+      set_tests_properties(${FULL_TEST_NAME} PROPERTIES ENVIRONMENT OMP_NUM_THREADS=${CURR_THREADS} PROCESSORS ${CURR_CORES} PROCESSOR_AFFINITY True)
+      if (ecut_DEP AND NOT ecut_DEP STREQUAL "${FULL_TEST_NAME}")
+        set_tests_properties(${FULL_TEST_NAME} PROPERTIES DEPENDS ${ecut_DEP})
+      endif()
+
+      if (ecut_LABELS)
+        set_tests_properties(${FULL_TEST_NAME} PROPERTIES LABELS "${ecut_LABELS}")
+      endif()
+
+      math(EXPR CURR_THREADS "${CURR_THREADS}+${THREAD_INCREMENT}")
+    endwhile()
+    set(CURR_THREADS ${THREAD_START})
+    math(EXPR CURR_RANKS "${CURR_RANKS}+${MPI_INCREMENT}")
+  endwhile()
+
+endfunction(EkatCreateUnitTest)
```

Comment:
This function will be very useful!

Reply:
It's mostly a copy+paste of CreateUnitTest from ScreamUtils.cmake (which now calls this one). I find it useful not just for the mpi/omp goodie at the end, but also cause writing

```
EkatCreateUnitTest(myexe "${sources}" INCLUDE_DIRS "${my_include_dirs}" LIBS "foo;bar" LABELS "basic;unit" LINKER_FLAGS "-ld" DEP "setup_test")
```
Is easier to parse for a reader than the sequence

```
add_executable(myexe ${sources})
target_link_libraries(myexe foo bar)
target_include_directories(myexe ${my_include_dirs})
target_link_options(myexe "-ld")
add_test (myexe_test myexe)
set_tests_properties(myexe_test PROPERTIES LABELS "basic;unit" DEPENDS setup_test)
```

The least satisfactory part for me is that it sets test dependencies of all the `A_npN_ompM` tests on a test `B`. But what if B was created with multiple thread/ranks options? There's no way for cmake to figure out the mpi/omp settings for that test, and since `B` is not a valid test name (only `B_npN_ompM` are), it will crap out. I guess in such a scenario, the user is out of luck, and needs to set dependencies manually on all the test, upon return from this function.